[Nguyen Anh Duc] iP

lilduckling opened the pull request:
# Avocado

> "English muffins with avocado is one of my favorite breakfasts." ~ Mia Hamm” ([source](https://www.azquotes.com/quotes/topics/avocados.html))

I don't know how the random quote is related. But hey, the chatbot is

- text-based
- easy to learn
- ~~FAST~~ ~~SUPER~~ EXTREMELY FAST to use

All you need to do is,

1. download it from [here](https://github.com/lilduckling/ip/releases).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you (_while eating avocado of course_)😉

And it is **FREE**!

Features:
 - [x] Managing tasks
 - [ ] Managing deadlines (coming soon)
 - [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

lilduckling committed:
Duke Level-0: Rename, Greet, Exit

lilduckling committed:
Duke Level-1: Echo

lilduckling committed:
Duke Level-1: Echo

lilduckling committed:
Duke Level-2: Add, List

lilduckling committed:
Duke Level-3: Mark as Done

lilduckling committed:
Duke Level-4: ToDo, Event, Deadline

lilduckling committed:
Duke A-TextUiTesting: Automated Text UI Testing

lilduckling committed:
Duke A-TextUiTesting: Automated Text UI Testing

lilduckling committed:
Duke Level-5: Handle Errors

lilduckling committed:
Duke Level-6: Delete

lilduckling committed:
Implement Level-7: Save tasks to file

lilduckling committed:
Merge Level-7: Save

lilduckling committed:
Implement Level-8

lilduckling committed:
Merge Level-8: Dates and Times

lilduckling committed:
Redesign Avocado with more OOP

lilduckling committed:
Merge A-MoreOOP: Improved OOP design

lilduckling committed:
Redesign into packages

lilduckling committed:
Merge A-Packages: Organize classes into packages

lilduckling committed:
Merge add-gradle-support into master

lilduckling committed:
Implement A-Gradle: Set up Gradle build automation

lilduckling committed:
Add A-JUnit

lilduckling committed:
Merge A-JUnit: Added JUnit tests for commands

lilduckling committed:
Add A-JUnit

lilduckling committed:
Fix build.gradle

lilduckling committed:
Implement A-JavaDoc: Add JavaDoc comments to Avocado chatbot

lilduckling committed:
Edit to follow coding standard

lilduckling committed:
Merge A-JavaDoc: Added JavaDoc documentation

lilduckling committed:
Merge A-CodingStandard: Applied coding standards

lilduckling committed:
Resolve merge conflicts in A-CodingStandard

lilduckling committed:
Implement Level-9: Add find feature

lilduckling committed:
Merge Level-9: Implement find feature

NoahKoh commented on src/main/java/avocado/Avocado.java:
```diff
@@ -0,0 +1,67 @@
+package avocado;
```
Package names are all in lowercase which is good. Applies to all other Java files

NoahKoh commented on src/main/java/avocado/Avocado.java:
```diff
@@ -0,0 +1,67 @@
+ * Handles initialization and execution of user commands.
+ */
+
+public class Avocado {
```
Class names are written in PascalCase, good. Applies to all other Java files as well

NoahKoh commented on src/main/java/avocado/command/Command.java:
```diff
@@ -0,0 +1,33 @@
+     *
+     * @return True if the command is an exit command, false otherwise.
+     */
+    public boolean isExit() {
```
Boolean methods sounds like boolean, good

NoahKoh commented on src/main/java/avocado/storage/Storage.java:
```diff
@@ -0,0 +1,138 @@
+ * Represents the storage of tasks in the hard disk.
+ */
+public class Storage {
+    private static final String FILE_PATH = "./data/tasks.txt";
```
static final variable name in uppercase, good

NoahKoh commented on src/main/java/avocado/task/TaskList.java:
```diff
@@ -0,0 +1,110 @@
+ * Represents a list of tasks.
+ */
+public class TaskList {
+    private final ArrayList<Task> tasks;    
```
Plural form used on collection variable, good

NoahKoh commented on src/main/java/avocado/Avocado.java:
```diff
@@ -0,0 +1,67 @@
+    private TaskList tasks;
+    private final Ui ui;
+
+    /**
+     * Constructs an Avocado instance with a specified storage file.
+     *
+     * @param filePath The path to the file used for saving/loading tasks.
+     */
```
Correct form for Javadoc comment, good

NoahKoh reviewed (commented):
Overall I find your code to be following the Java coding standards, no issue. LGTM

lilduckling committed:
Implement Level-10:GUI

lilduckling committed:
Merge Level-10: GUI

Wailydest commented on src/main/java/avocado/parser/Parser.java:
```diff
@@ -0,0 +1,61 @@
+        } else if (fullCommand.equalsIgnoreCase("list")) {
+            return new ListCommand();
+        } else if (fullCommand.startsWith("mark ")) {
+            return new MarkCommand(Integer.parseInt(fullCommand.split(" ")[1]) - 1);
```
Integer.parseInt() may throw a NumberFormatException if the user doesn't properly specify an integer. Shouldn't that be handled?

Wailydest commented on src/main/java/avocado/storage/Storage.java:
```diff
@@ -0,0 +1,138 @@
+                    return null;
+            }
+        } catch (Exception e) {
+            System.out.println("Skipping corrupted line: " + line);
```
It's a nice way to handle corrupted data. Great job!

Wailydest commented on src/main/java/avocado/storage/Storage.java:
```diff
@@ -0,0 +1,138 @@
+            boolean isDone = parts[1].equals("1");
+            String description = parts[2];
+
+            switch (type) {
```
Shouldn't switch expression follow the style described [here](https://se-education.org/guides/conventions/java/intermediate.html)?
```Java
switch (type) {
case "X":
    statements;
    break;
case "Y":
    statements;
    break;
}
```

Wailydest commented on src/main/java/avocado/Avocado.java:
```diff
@@ -0,0 +1,80 @@
+            try {
+                String fullCommand = ui.readCommand();
+                ui.showLine();
+                Command command = Parser.parse(fullCommand);
```
That's a very OOP way to do this. Good job!

Wailydest reviewed (commented):
Overall, LGTM. The code has followed a clear OOP structure, exhibiting key aspects such as encapsulation and inheritance. The code is well-structured and readable. JavaDoc was widely and consistently applied, which made the code pleasant to read. The tests were quite nice and comprehensive. Keep up the nice work!

lilduckling committed:
Implement A-Assertions: Add assertions for critical assumptions

- Added assertions in TaskList.java to validate task index before marking tasks as done.

- Added assertions in Avocado.java to validate task index before marking tasks as done.

- Added assertions in Launcher.java to validate task index before marking tasks as done.

These assertions improve code reliability and maintainability while ensuring key assumptions hold.

lilduckling committed:
Refactor code to improve readability

- Reduce all long method
- Fix deep nested code in Parser.java
- Add a Constant.java
- Removed unused imports and cleaned up the codebase.
- Ensured consistent indentation, spacing, and formatting.

These improvements enhance code readability thus code quality

lilduckling committed:
Merge pull request #1 from lilduckling/branch-A-Assertions

Implement A-Assertions: Add assertions for critical assumptions

lilduckling committed:
Merge pull request #2 from lilduckling/branch-A-CodeQuality

Refactor code to improve readability

lilduckling committed:
Implement C-Tagging: Allow tasks to store tags directly

- Allow add and delete tags to a tasks
- Allow get all tasks from a tags

This change makes it easier to categorise and search for tasks

lilduckling committed:
Merge C-Tagging: Added tagging support with HashMap

lilduckling committed:
Minor debug and add documentaries

lilduckling committed:
Fix ReadMe file

- Update java version
- Add new way to run the program

This make the file more detailed for users

lilduckling committed:
Clear data file

- Clear data file in data/tasks.txt

This is to delete unreal tasks pre-added

lilduckling committed:
Update javaDoc

- Add comments for more java file in src

lilduckling committed:
Update javaDoc

- Add comments for more java file in src

lilduckling committed:
Fix ReadMe file

- Add Ui.png in ReadMe

This make the file more detailed for users
